### PR TITLE
GC tests where summarizer loads from an older snapshot

### PR DESCRIPTION
**Pull request by agarwal-navin**
Added tests which repros a bug where GC data is not correctly update on refreshLatestSummaryFromSnapshot:

refreshLatestSummaryFromSnapshot downloads a specific snapshot from the server and various latest summary state is updated from this snapshot. However, the GC data in summarizer nodes is not updated from this snapshot which can lead to incorrect behavior. For example, consider the following scenarios:
A data store (say ds1) is unreferenced in summary1 at seq#100.
An op references ds1 and it is referenced in summary2 at seq#101.
A new summarizer loads from summary1 (seq#100). While catching up to the latest, gets ack for summary2 and calls refreshLatestSummary. The referenceSequenceNumber for ds1 is updated to seq#101.
In the next summary (summary3), if ds1 did not change, it will not run GC again since its referenceSequenceNumber is the same since last summary (summary2).
GC will use data from previous summary which is the data from summary1. This is because GC did not update data from summary2. In this summary, ds1 will be unreferenced incorrectly.

Basically, any data stores that did not change between refreshLatestSummaryFromSnapshot and the next summary may have incorrect GC data.

Demonstrates the bug in https://dev.azure.com/fluidframework/internal/_workitems/edit/2821.

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        containerRuntime = dataStoreA._context.containerRuntime as IContainerRuntime;
+        await waitForContainerConnection(mainContainer);
+    });
+
+    it("updates referenced nodes correctly when loading from an older summary", async () => {
+        const summarizer1 = await createSummarizer(provider, mainContainer);
+
+        // Create a data store and mark it unreferenced to begin with.
+        const dataStoreB = await requestFluidObject<ITestDataObject>(
+            await containerRuntime.createDataStore(TestDataObjectType), "");
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts:87]: Can you just get the handle of the datastore here and call get?

File: packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    /**
+     * Submits a summary and returns the reference state for all the nodes in the container.
+     * If a node is referenced, its value is true. If it's unreferenced, its value is false.
+     * @returns a map of nodePath to its unreferenced timestamp.
+     */
+    async function getReferenceState(summaryTree: ISummaryTree) {
+        const gcState = getGCStateFromSummary(summaryTree);
+        assert(gcState !== undefined, "GC tree is not available in the summary");
+
+        const nodeReferenceStates: Map<string, boolean> = new Map();
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts:42]: Thought: `nodeIsReferencedMap`? and then the if statement below could be
`nodeIsReferencedMap.set(nodePath.slice(1), nodeData.unreferencedTimestampMs === undefined ? true: false);`

**Review by tyler-cai-microsoft — APPROVED** [review]
LGTM, just had some minor comments

File: packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts:87]: Did the change go in where createDataStore returns a handle? If so, yes we can do that.

Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts:87]: I think so

Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcLoadingFromOlderSummaries.spec.ts]: Changed it.
